src/wrap_gradient8.c

```diff
@@ -0,0 +1,29 @@
+#include <stddef.h>
+#include <stdio.h>
+#include <stdint.h>
+#include "topotoolbox.h"
+
+#if defined(_OPENMP)
+#include <omp.h>
+#endif
+
+//    This is a working version to run gradient8
+//
+void wrap_gradient8(float *outputR, float *demR, float *cellsizeR,int *use_mpR, ptrdiff_t *dimsR){ //might be usefull to not ask for a ptrdiff_t in the first place
+                                 //and let the user input a long array
```

Comment:
can you look into my fillsink branch and tell me, if you liked the way i did it there?

Reply:
Okay if i see my mistake correctly then i kind of did it in a way where i hand over the wrong data type and "praying" that nothing changes in the memory adresse, so that afterwards it gets picked up correctly and reconverted to an integer.  And it worked, but i corrected it. 

For some reason, when i was fiddling around it nether worked direktly, nor printed out the contents. That's why i started those weird conversion. Same as the one with the long cast, which i did, because the numbers where always off, like it tried to convert more "bits" than int is and i've read that ptrdiff_t always takes in long.

Reply:
I think I understand what was happening here, and it is good to understand it for the future.

When you pass a pointer to a particular type, C interprets the memory address as containing whatever type you give it. So when you dereference a pointer like so

```C
ptrdiff_t dims[2] = {100,200}; // C arrays are essentially the same as pointers.

// Dereference the pointer to extract the value pointed to by dims
ptrdiff_t rows = *dims; // rows = 100
```

C reads 8 bytes of memory because `ptrdiff_t` is usually 8 bytes on a 64-bit platform, from the address given by `dims `and interprets it as a `ptrdiff_t`. If you assign the `dims` pointer to another type and dereference it like so

```C
ptrdiff_t dims[2] = {100, 200};

int *idims = dims;

int rows = *idims;
```

this is undefined behavior and the value actually assigned to `rows` 

Something similar was happening when you had the signature `wrap_gradient8(float *outputR, float *demR, float *cellsizeR, int *use_mpR, ptrdiff_t *dimsR)`. R was passing an `int *` to the `dimsR` argument, but you were interpreting it as a `ptrdiff_t *`. If you dereferenced that pointer, you would most likely get garbage out. But you ended up converting it back `int *pdims = dimsR;` and then when you dereferenced it, you happened to get the right answer. I think this would still be undefined behavior, so the compiler is not guaranteed to give you the right answer if you do it like that.

The correct solution as I mentioned is to have `wrap_gradient8` take `int *dimsR` and then convert the `int`s to `ptrdiff_t`s:

```C
wrap_gradient8(float *outputR, float *demR, float *cellsizeR, int *use_mpR, int *dimsR) {

    // The conversion to ptrdiff_t will happen implicitly.
    ptrdiff_t dims[2] = {dimsR[0], dimsR[1]}; 

    // ... call gradient8, etc.

}

```

Reply:
I should also mention that when I compile your original code, I get a compiler warning

```
wrap_gradient8.c:16:17:warning: initialization of ‘int *from incompatible pointer type ‘ptrdiff_t *{aka ‘long int *-Wincompatible-pointer-types
      16 |    int *pdims = dimsR
```
which highlights the incorrect conversion that I described above. It's generally a good idea to pay attention to these warnings. I will look into whether we can make the compiler throw an error rather than just warning you about these kinds of problems.

Reply:
Done. 

About these warnings and to justify myself. I am aware of them and tend to ignore them because, for me, they are time consuming to fix and i was more focused on getting the code to run some what. So, when i know if and how everthing works then i can refine the code and make it secure. 

But in the end, it might be more important to sort these warings out before going on, even if i spent time on fixing something that won't be used in the end. So, i can definitely start to focus and address these compiler warnings.